### PR TITLE
Isaac Del Rosario -- Carets

### PR DESCRIPTION
**Pull request by hisaacdelr**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |    I had to make BlockReservation its own class that has-a Reservation instead of making BlockReservation a sub-class of Reservation as I originally intended and implemented. This was because I couldn't get each room's availability instance variable to update when checking its availability for future dates. I decided that availability would be determined by if it was attached to a Reservation which had check_in and check_out instance variables I could check for a single or range of date's availability. |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  Although I didn't implement it eloquently, I gained a better hold on has-a/is-a relationships and their importance when designing classes that interact with each other. Admittedly, as I got further down in the waves and requirements changed, I realized my code got less flexible to work with as I added more and more features/checks into each method which Metz warned about in the reading.   |
| Describe a _nominal test_ that you wrote for this assignment. |  In my admin.reserve method that creates single-room-reservations, I test if the check-in and check-out overlap with a prior reservation when they're both within the date range of the prior reservation.   |
| Describe an _edge case test_ that you wrote for this assignment. |  Again with my .reserve method, an edge case is the constraint that you can reserve a room when it falls on the check-out date of another (block) reservation.   |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |  I prefer using this approach, it felt easier to list out a mini to-do list before having to think about syntax. But after getting engrossed in my code and focusing on making it just work, admittedly, following conventions and design concepts became a second or third priority... |

### CODE REVIEW

File: lib/room.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    end
+
+
+    def available_at?(check_date)
+      # because using the ! would return nil if all values were unique
+      return !(@booked_dates.include?(check_date))
+    end
+
+    def set_booked_dates(admin)
+      admin.reservations.each do |reservation|
```
Comment (CheezItMan) [inline lib/room.rb:31]: This is an example of __tight coupling__.   The Room class, needs to know that `admin` has reservations and that it's an array.

It's also a mix of purposes, why is a room setting reservations in the admin?  Why wouldn't the hotel set the reservation, or have the room set it's reservation?

It also seems like this room goes through all of admin's reservations and adds them all at once.

File: lib/room.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    def initialize(room_number)
+      # @blocked_off = false #for Wave 3
+      @cost = 200
+      @room_number = room_number
+      @booked_dates = []
+    end
+
+    # Returns 20 room objects as an array
+    def self.all
```
Comment (CheezItMan) [inline lib/room.rb:16]: This will make a new set of 20 rooms each time, even if it already exists...

File: lib/admin.rb [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      @block_reservations.each do |other_block_reservation|
+        other_block_reservation_room_numbers = other_block_reservation.rooms.map{|room| room.room_number}
+        if other_block_reservation.overlap?(check_in, check_out) && other_block_reservation_room_numbers.include?(room_num) # the last conditional means if Array1 & Array2 have elements in common
+          raise ArgumentError.new("There's overlap with this block reservation and an existing block reservation's date")
+        end
+      end
+
+      #Check all reservations if it can be made
+      @reservations.each do |reservation|
+        if reservation.room.room_number == room_num && reservation.overlap?(check_in, check_out)
```
Comment (CheezItMan) [inline lib/admin.rb:28]: same issues as above.

File: lib/admin.rb [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    def reservations_at_date(date)
+      # returns an array of reservations that have the given date in their range
+      list_of_reservations_at_date = @reservations.select do |reservation|
+        reservation.check_in <= date && reservation.check_out > date
+      end
+      return list_of_reservations_at_date
+    end
+
+    # As an administrator, I can get the total cost for a given reservation
+    def total_cost(reservation)
```
Comment (CheezItMan) [inline lib/admin.rb]: Why is the admin figuring the cost  of a specific reservation?  

This is an issue with separation of concerns.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Good number of commits and good messages.  |
Answer comprehension questions	| Check, I'm glad you see where your code-first mentality lead you.  |
**Design** |
Each class is responsible for a single piece of the program |  You've got some separation of concerns issues here.  I put in some comments in your code.  Essentially `Admin` is reaching into the internals of Reservation and Block to check things it should simply go through instance methods to do.  |
Classes are loosely coupled |  There's some tight coupling here, coming from the issues above.  I left some comments in your code.  |
**Wave 1** |
List rooms |  Check  |
Reserve a room for a given date range |  Check  |
List reservations for a given date | Check  |
Calculate reservation price |  Check, but issues with separation of concerns. |
Invalid date range produces an error |  Check, good use of custom errors. |
**Wave 2** |
View available rooms for a given date range | Check |
Reserving a room that is not available produces an error |  Check |
**Wave 3** | 
Create a block of rooms | Check  |
Check if a block has rooms |  Check  |
Reserve a room from a block |  Looks like you're not quite finished yet. |
Test coverage |  100%  |
**Additional Feedback** |  There's a lot of good here, custom errors, business logic in `Room` and `Reservation`, but you're not using them too much in `Admin`.  There's still work to be done in separating concerns.  I left some comments in your code to reference.  I hope a lot was learned in this project.    |

**Review by CheezItMan — COMMENTED** [review]
Feedback for some specific methods.
